apa.csl

```diff
@@ -50,94 +51,296 @@
+            <label form="short" prefix=", " text-case="title"/>
+          </names>
+          <names variable="editor translator" delimiter="; ">
+            <name and="symbol" initialize-with=". " delimiter=", "/>
+            <label form="short" prefix=", " text-case="title"/>
+          </names>
+        </group>
+      </else>
     </choose>
   </macro>
   <macro name="author">
-    <names variable="author">
-      <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
-      <label form="short" prefix=" (" suffix=")" text-case="capitalize-first"/>
-      <substitute>
-        <names variable="editor"/>
-        <names variable="translator"/>
-        <choose>
-          <if type="report">
-            <text variable="publisher"/>
-            <text macro="title"/>
-          </if>
-          <else>
+    <choose>
+      <if type="song">
+        <names variable="composer" delimiter=", ">
+          <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
+          <substitute>
+            <names variable="original-author"/>
+            <names variable="author"/>
+            <names variable="translator">
+              <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
+              <label form="short" prefix=" (" suffix=")" text-case="capitalize-first"/>
+            </names>
+            <choose>
+              <if variable="container-title">
+                <text variable="title" form="short" quotes="true"/>
+              </if>
+              <else>
+                <text variable="title" form="short" font-style="italic"/>
+              </else>
+            </choose>
+            <text macro="format"/>
+          </substitute>
+        </names>
+      </if>
+      <else-if type="treaty"/>
+      <else>
+        <names variable="author" delimiter=", ">
+          <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
+          <substitute>
+            <names variable="illustrator"/>
+            <names variable="composer"/>
+            <names variable="director">
+              <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
+              <label form="long" prefix=" (" suffix=")" text-case="capitalize-first"/>
+            </names>
+            <choose>
+              <if variable="container-title">
+                <names variable="translator">
+                  <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
+                    <label form="short" prefix=" (" suffix=")" text-case="capitalize-first"/>
+                    <substitute>
```

Comment:
The thought here was this: If there is a container-title and no author, then it's reasonable to assume that the "translator" refers to the translator of the chapter, rather than the container as a whole (i.e., the translator of a chapter in an edited book). So what this `choose` is doing is changing the relative order of translator and editor depending on if there is a container title or not. I can see a way to streamline this though. 

Two questions:
1. Is there a way to test for `editortranslator` (the editor and translator being the same)? The goal being to show the `editortranslator` label if applicable, but then to show only one of `editor` or `translator` if it's not.

1. Currently, if the `label` formatting is changed during a `names` element (e.g., in the current code, it starts with no label for author so that illustrator and composer will be shown without labels, then does long labels for director, then switches to short labels for editor/translator), the label formatting always conforms to the last label format it encounters. So, for example, if I have:
```<names variable="author" delimiter=", ">
          <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
          <substitute>
            <names variable="illustrator"/>
            <names variable="composer"/>
            <names variable="director">
              <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
              <label form="long" prefix=" (" suffix=")" text-case="capitalize-first"/>
            </names>
            <names variable="editor">
              <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
              <label form="short" prefix=" (" suffix=")" text-case="capitalize-first"/>
            </names>
            <names variable="editorial-director"/>
            <names variable="translator"/>
          </substitute>
</names>
```
then editorial-director and translator will be rendered by citeproc-js with short labels (rather than no labels). This seems like it isn't the correct behavior, correct? To get short labels with editorial-director and translator, I should need to use a complete names-name-label construct for each, correct?

Reply:
Also, do you have a preference for either of these?

1. No label given for `author`, but then a label specified for each of `director`, `translator` (conditional), `editor`, `editorial-director`, and `translator` (6 name lines):
```
<names variable="author" delimiter=", ">
  <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
  <substitute>
    <names variable="illustrator"/>
    <names variable="composer"/>
    <names variable="director">
      <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
      <label form="long" prefix=" (" suffix=")" text-case="capitalize-first"/>
    </names>
    <choose>
      <if variable="container-title">
        <names variable="translator">
          <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
            <label form="short" prefix=" (" suffix=")" text-case="capitalize-first"/>
        </names>
      </if>
    </choose>
    <names variable="editor">
      <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
        <label form="short" prefix=" (" suffix=")" text-case="capitalize-first"/>
    </names>
    <names variable="editorial-director">
      <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
        <label form="short" prefix=" (" suffix=")" text-case="capitalize-first"/>
    </names>
    <names variable="translator">
      <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
        <label form="short" prefix=" (" suffix=")" text-case="capitalize-first"/>
    </names>
    <choose>
      <if type="report">
        <text variable="publisher"/>
      </if>
    </choose>
    <text macro="title"/>
    <text macro="format"/>
  </substitute>
</names>
```

2. Short label given for `author`, but then a different label specified for `illustrator` (empty), `composer` (empty), and `director` (long) (4 name lines):
```
<names variable="author" delimiter=", ">
  <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
  <label form="short" prefix=" (" suffix=")" text-case="capitalize-first"/>
  <substitute>
    <names variable="illustrator">
      <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
    </names>
    <names variable="composer">
      <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
    </names>
    <names variable="director">
      <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
      <label form="long" prefix=" (" suffix=")" text-case="capitalize-first"/>
    </names>
    <choose>
      <if variable="container-title">
        <names variable="translator"/>
      </if>
    </choose>
    <names variable="editor"/>
    <names variable="editorial-director"/>
    <names variable="translator"/>
    <choose>
      <if type="report">
        <text variable="publisher"/>
      </if>
    </choose>
    <text macro="title"/>
    <text macro="format"/>
  </substitute>
</names>
```

The second is somewhat shorter, but also less readable.